dozenal/doc/doman.1
." +AMDG
." Process with:
." groff -man -Tascii doman.1
.TH doman 1 "September 2017" Linux "User Manuals"
.SH NAME
doman \- convert dozenal integers to arbitrary non-place-value systems
.SH SYNOPSIS
.B doman [-V] [-R 
.I dozenal integer
.B ] [-r
.I ranks
.B ] [-s
.I symbols
.B ]
.SH DESCRIPTION
.B doman
converts dozenal integers to arbitrary non-place-value
systems.  Only accept dozenal integers as input; fractional
places will be silently truncated.  Can accept arbitrary
symbols; e.g., to duplicate (non-subtractive) Roman numerals, use 
\fB-s"M,D,C,L,X,V,I" -r"6E4,358,84,42,X,5,1"\fR.  Any character
or string of characters may be used as a symbol.
.SH OPTIONS
.TP
.BR -V
Print program and version information and exit successfully.
.TP
.BR "-R integer"
The number you wish \fBdoman\fR to convert.
.TP
.BR "-r ranks"
Takes as a comma-separated list the ranks you wish to use
for assembling the non-place-value number.  E.g., for
(non-subtractive) Roman numerals, you'd use
"6E4,358,84,42,X,5,1", those being the numbers that have
separate symbols in Roman numerals.  You can have as many or
as few of these as you wish.
.PP
.RS
Notice that this is a comma-separated list of \fIdozenal\fR
values; decimal values will be interpreted as dozenal,
which, if you meant to use decimal values, will not give you
the result you want.
.PP
Note that there must be the same number of ranks as symbols;
if there isn't, \fBdoman\fR will emit an error and die.
.PP
By default, \fBdoman\fR sets this equal to:
\fB1728,1584,1440,864,720,576,144,132,120,72,60,48,12,11,10,6,5,4,1\fR.
This produces dozenal Roman numerals, of the sort discussed
in several dozenal publications.
.RE
.TP
.BR "-s symbols"
Takes as a comma-separated list the symbols you wish to use
for each rank.  E.g., for (non-subtractive) Roman numerals,
you'd use "M,D,C,L,X,V,I".  However, any symbols can be
used, of whatever length; multi-byte characters (or multiple
characters) are fine.
.PP
.RS
Note that there must be the same number of ranks as symbols;
if there isn't, \fBdoman\fR will emit an error and die.
.PP
By default, \fBdoman\fR defines this as follows:
\fBM,CM,CCM,D,CD,CCD,C,XC,XXC,L,XL,XXL,X,IX,IIX,V,IV,IIV,I\fR.
This produces dozenal Roman numerals, of the type discussed
in several dozenal publications.
.RE
.SH SUBTRACTIVE NOTATION
.PP
Roman numerals include something called \fIsubtractive
notation\fR (though Romans themselves didn't use it very
often, and when they did they very rarely used it
consistently).  By this means, a lower-ranked symbol may be
placed prior to a higher-ranked symbol to indicate
subtraction.  E.g., \fBIV\fR would means \fB5 - 1\fR, or 4.
.PP
\fBdoman\fR does have the ability to do this subtractive
notation; however, it can't do so in a general way.  To
accomplish this, include the subtractive numbers as ranks,
and their compound symbols.  E.g., to make \fBIV\fR work as
intended:
.RS
.B -s"V,IV,I" -r"5,4,1"
.RE
That should do the trick.  To completely duplicate the rules
of conventional Roman numerals, try the following:
.RS
.nf
.B -r"6E4,630,358,294,84,76,42,34,X,9,5,4,1" -s"M,CM,D,CD,C,XC,L,XL,X,IX,V,IV,I"
.fi
.RE
.SH BUGS
None known at this time.
.SH AUTHOR
Donald P. Goodman III <dgoodmaniii at gmail dot com>
.SH "SEE ALSO"
.BR doz (1),
.BR dozdc (1),
.BR tgmconv (1),
.BR dozdate (1),
.BR dozcal (1),
.BR dozword (1),
.BR doman (1),
.BR dozstring (1),
.BR dozpret (1)


dozenal/doman/doman.c
/* +AMDG */
/*
 * This document was begun on 15 Aug 1201, the feast of St.
 * Hyacinth of Poland, C, and it is humbly dedicated to him and
 * to the Immaculate Heart of Mary for her prayers, and to
 * the Sacred Heart of Jesus, for His mercy.
 */

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<unistd.h>
#include<getopt.h>
#include"errors.h"
#include"conv.h"

#define MAXLEN 256

int *ranks;
char **lets;

int change_vals(int *val,int place);
int check_args();
int count_commas(char *s);

int main(int argc, char **argv)
{
	int ex = 7584;
	int mod;
	int i, j;
	char answer[MAXLEN] = "";
	char c;
	int numranks;
	char *token;
	/* initialize the default ranks and lets */
	ranks = (int *)malloc(20 * sizeof(int));
	ranks[0] = 1728; ranks[1] = 1584; ranks[2] = 1440;
	ranks[3] = 864; ranks[4] = 720; ranks[5] = 576;
	ranks[6] = 144; ranks[7] = 132; ranks[8] = 120;
	ranks[9] = 72; ranks[10] = 60; ranks[11] = 48;
	ranks[12] = 12; ranks[13] = 11; ranks[14] = 10;
	ranks[15] = 6; ranks[16] = 5; ranks[17] = 4;
	ranks[18] = 1; ranks[19] = 0;
	lets = malloc(20 * sizeof(char *));
	for (i = 0; i < 20; ++i)
		*(lets+i) = malloc(4 * sizeof(char));
	strcpy(lets[0],"M"); strcpy(lets[1],"CM"); strcpy(lets[2],"CCM");
	strcpy(lets[3],"D"); strcpy(lets[4],"CD"); strcpy(lets[5],"CCD");
	strcpy(lets[6],"C"); strcpy(lets[7],"XC"); strcpy(lets[8],"XXC");
	strcpy(lets[9],"L"); strcpy(lets[10],"XL"); strcpy(lets[11],"XXL");
	strcpy(lets[12],"X"); strcpy(lets[13],"IX"); strcpy(lets[14],"IIX");
	strcpy(lets[15],"V"); strcpy(lets[16],"IV"); strcpy(lets[17],"IIV");
	strcpy(lets[18],"I"); strcpy(lets[19],"\0");
		
	opterr = 0;
	while ((c = getopt(argc,argv,"m:R:r:s:V")) != -1) {
		switch(c) {
		case 'V':
			printf("doman v1.0\n");
			printf("Copyright (C) 1201 (2017) by Donald P.  Goodman III\n");
			printf("License GPLv3+:  GNU GPL Version 3 or later "
				"<http://gnu.org/licenses/gpl.html>\n");
			printf("This is free software; you are free to change "
				"and redistribute it.  There is NO WARRANTY, "
				"to the extent permitted by law.\n");
			exit(SUCCESS);
			break;
		case 'R':
			ex = (int)doztodec(optarg);
			break;
		case 'r':
			numranks = count_commas(optarg) + 2;
			ranks = (int *)realloc(ranks,numranks * sizeof(int));
			i = 0;
			token = strtok(optarg,",");
			while (token != NULL) {
				ranks[i++] = (int)doztodec(token);
				token = strtok(NULL,",");
			}
			ranks[i] = 0;
			break;
		case 's':
			numranks = count_commas(optarg) + 2;
			for (i = 0; strcmp(lets[i],"\0"); ++i)
				free(lets[i]);
			free(lets[i]);
			lets = realloc(lets,numranks * sizeof(char *));
			i = 0;
			token = strtok(optarg,",");
			while (token != NULL) {
				*(lets+i) = malloc((strlen(token) + 1) * sizeof(char));
				strcpy(*(lets+i),token);
				token = strtok(NULL,",");
				i++;
			}
			*(lets+i) = malloc(2 * sizeof(char));
			strcpy(*(lets+i),"\0");
			break;
		case '?':
			if ((optopt == 'R') ||
				(optopt == 'm') || (optopt == 's')) {
				fprintf(stderr,"doman:  option \"%c\" requires "
					"an argument\n",optopt);
				exit(OPT_REQ_ARG);
			}
			fprintf(stderr,"doman:  unrecognized option "
				"\"%c\"\n",optopt);
			exit(UNREC_OPT);
			break;
		}
	}
	check_args();
	/* here's the action */
	for (i = 0; ranks[i] != 0; ++i) {
		mod = change_vals(&ex,ranks[i]);
		for (j = 0; j < mod; ++j)
			strcat(answer,lets[i]);
	}
	printf("%s\n",answer);
	free(ranks);
	for (i = 0; strcmp(lets[i],"\0"); ++i)
		free(lets[i]);
	free(lets[i]);
	free(lets);
	return 0;
}

int change_vals(int *val, int place)
{
	int mod;

	mod = *val / place;
	*val %= place;
	return mod;
}

int check_args()
{
	int i, j;
	for (i = 0; ranks[i] != 0; ++i);
	for (j = 0; strlen(lets[j]) != 0; ++j);
	if (i != j) {
		fprintf(stderr,"doman:  error:  you have %d "
			"ranks and %d symbols; you should have "
			"an equal number of both\n",i,j);
		exit(MISMATCHED_RANKS_SYMS);
	}
	return 0;
}

int count_commas(char *s)
{
	int i;
	int j = 0;

	for (i = 0; s[i] != '\0'; ++i) {
		if (s[i] == ',')
			++j;
	}
	return j;
}
